### PR TITLE
[enh] Add avx2 support in finiteness_checker

### PR DESCRIPTION
**Pull request by icfaust**
# Description
Duplicate AVX512 functions for AVX2 by switching certain numbers to half size and changing instructions from 512 to 256 bit width.  Due to the hardcoded nature of the functions, it is not easily templated out without performance loss. This implementation should improve sklearnex performance on standard benchmarks.

Changes proposed in this pull request:
- Add avx2 finite sum check
- Add avx2 finiteness per element check
- Add avx2 SOA supports

Tasks
- [x] Implement AVX2
- [ ] Get it to compile
- [ ] Green CI
- [ ] Run sklearnex Benchmarks

### CODE REVIEW

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 1]
```diff
@@ -192,6 +192,30 @@ DataType sumWithAVX512(size_t n, const DataType * dataPtr)
     return sum;
 }
 
+template <typename DataType>
+DataType sumWithAVX2(size_t n, const DataType * dataPtr)
+{
+    const size_t nPerInstr = 32 / sizeof(DataType);
+    DataType sum;
+    if (sizeof(DataType) == 4)
```
Comment (icfaust) [inline cpp/daal/src/data_management/finiteness_checker.cpp:200]: Am I missing something about how the compiler works, or why is this branching necessary??

Comment (icfaust) [reply cpp/daal/src/data_management/finiteness_checker.cpp:200]: Now I understand

**Comment by icfaust** [comment]
test fail related to rbf kernel, which doesn't use this code

**Review by Alexandr-Solovev — COMMENTED** [review]
In general it looks good to me, but, it's better to change the order of the new functions. Like firstly avx2 then avx512. Also, please use onedal-benchmarks with the kernel profiler as well as intelex benchmarks to check what kernels have been improved. I would like to say that you will get similar performance benefits in both.

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 2]
```diff
@@ -76,6 +57,7 @@ bool valuesAreNotFinite(const double * dataPtr, size_t n, bool allowNaN)
 template <typename DataType, daal::CpuType cpu>
 DataType computeSum(size_t nDataPtrs, size_t nElementsPerPtr, const DataType ** dataPtrs)
 {
+    std::cout << "standard sum \n";
```
Comment (samir-nasibli) [inline cpp/daal/src/data_management/finiteness_checker.cpp:60]: remove some debug prints

File: cpp/daal/src/data_management/finiteness_checker_cpu.cpp [hunk 3]
```diff
@@ -0,0 +1,917 @@
...
+
+            __mmask16 expAreOnes   = _mm512_cmpeq_epi32_mask(exp512Mask, expBits);
+            __mmask16 fracAreZeros = _mm512_cmpeq_epi32_mask(zero512, fracBits);
+
+            // "values aren't finite" = "exponent bits are ones" AND ( "fraction bits are zeros" OR NOT "NaN is allowed" )
+            __mmask16 orMask    = _kor_mask16(fracAreZeros, notAllowNaNMask);
+            __mmask16 finalMask = _kand_mask16(expAreOnes, orMask);
+            endMask             = _kor_mask16(endMask, finalMask);
+        }
+        if (_cvtmask16_u32(endMask) != 0) notFinitePtr[iBlock] = true;
```
Comment (icfaust) [inline cpp/daal/src/data_management/finiteness_checker_cpu.cpp:686]: Note to reviewers: I have changed these lines to remove branching in the for loop. This was otherwise hidden due to the movement of the file, but otherwise the AVX512 portion is unchanged.

**Comment by icfaust** [comment]
Private CI just shows an unrelated issue with LibLinear convergence issues, which shouldn't be touched by this code / is likely sporadic.

**Comment by icfaust** [comment]
private CI run with intel/scikit-learn-intelex#1759 http://intel-ci.intel.com/eedfc7b0-6419-f133-b20a-a4bf010d0e2e

**Comment by icfaust** [comment]
wrote a special sklearnex version which will print warnings when an inf or nan is observed, which should show up in pytest for sklearn. https://github.com/intel/scikit-learn-intelex/compare/main...icfaust:scikit-learn-intelex:test/warning_finite?expand=1

I have run a special private CI run with this branch to see if it shows up at all in sklearn conformance tests, to see how much sklearn tests _assert_all_finite, specifically to see if it is actually activated: http://intel-ci.intel.com/eee83e31-ce3d-f14a-985d-a4bf010d0e2e

**Comment by icfaust** [comment]
> wrote a special sklearnex version which will print warnings when an inf or nan is observed, which should show up in pytest for sklearn. https://github.com/intel/scikit-learn-intelex/compare/main...icfaust:scikit-learn-intelex:test/warning_finite?expand=1
> 
> I have run a special private CI run with this branch to see if it shows up at all in sklearn conformance tests, to see how much sklearn tests _assert_all_finite, specifically to see if it is actually activated: http://intel-ci.intel.com/eeea9220-3038-f1df-9c06-a4bf010d0e2e (running against onedal-src/oneDAL/main)

So after activating a warning for when an inf or nan is spotted and by running all sklearn conformance testing, the only time an inf or nan occurs in the sklearn testing is here: https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/tests/test_validation.py#L981 Therefore intel/scikit-learn-intelex#1759 is sadly necessary

File: cpp/daal/src/data_management/finiteness_checker.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/** file finiteness_checker.h */
+/*******************************************************************************
+* Copyright 2024 Intel Corporation
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker.h:3]: Please align copyrights with the new UXL requirements. 
```suggestion
* Copyright contributors to the oneDAL project
```

File: cpp/daal/src/data_management/finiteness_checker.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+{
+using namespace daal::internal;
+
+typedef daal::data_management::NumericTable::StorageLayout NTLayout;
+
+const uint32_t floatExpMask  = 0x7f800000u;
+const uint32_t floatFracMask = 0x007fffffu;
+const uint32_t floatZeroBits = 0x00000000u;
+
+const uint64_t doubleExpMask  = 0x7ff0000000000000uLL;
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker.h:37]: `uLL` at the end looks strange. Please choose either 'ull' or 'ULL'.

File: cpp/daal/src/data_management/finiteness_checker_cpu.cpp [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+        *((uint64_t *)(&inf)) = doubleExpMask;
+    return inf;
+}
+
+    #if defined(__AVX2__) and !defined(__AVX512F__)
+
+template <typename DataType>
+DataType sumWithAVX2(size_t n, const DataType * dataPtr)
+{
+    const size_t nPerInstr = 32 / sizeof(DataType);
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_cpu.cpp:147]: Those calculations are not obvious.
It's better to have:
```suggestion
    constexpr size_t avx2RegisterLength = 256;
    constexpr size_t numberOfBitsInByte = 8;
    constexpr size_t nPerInstr = avx2RegisterLength / (numberOfBitsInByte * sizeof(DataType));
```

File: cpp/daal/src/data_management/finiteness_checker_cpu.cpp [hunk 7]
```diff
@@ -0,0 +1,917 @@
...
+        size_t ptrIdx        = iBlock / nBlocksPerPtr;
+        size_t blockIdxInPtr = iBlock - nBlocksPerPtr * ptrIdx;
+        size_t start         = blockIdxInPtr * nPerBlock;
+        size_t end           = blockIdxInPtr == nBlocksPerPtr - 1 ? start + nPerBlock + nSurplus : start + nPerBlock;
+
+        pSums[iBlock] = sumWithAVX2<DataType>(end - start, dataPtrs[ptrIdx] + start);
+    });
+
+    return sumWithAVX2<DataType>(nTotalBlocks, pSums);
+}
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_cpu.cpp:210]: This code is almost 100% copy-paste from AVX512 code.
I think it can be unified if sumWithXXX and computeSumXXXImpl functions would have a `cpu` template parameter.

File: cpp/daal/src/data_management/finiteness_checker_cpu.cpp [hunk 8]
```diff
@@ -0,0 +1,917 @@
...
+    return computeSumAVX2Impl<float>(nDataPtrs, nElementsPerPtr, dataPtrs);
+}
+
+template <>
+double computeSum<double, avx2>(size_t nDataPtrs, size_t nElementsPerPtr, const double ** dataPtrs)
+{
+    return computeSumAVX2Impl<double>(nDataPtrs, nElementsPerPtr, dataPtrs);
+}
+
+double computeSumSOAAVX2Impl(NumericTable & table, bool & sumIsFinite, services::Status & st)
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_cpu.cpp:224]: This is also a big copy-paste. I think there should be ways to avoid it.

**Comment by icfaust** [comment]
This will likely pass CI, but performance benchmarks are necessary due to the underlying changes in the CPU function dispatching.

**Comment by icfaust** [comment]
Things required before re-review: a privateCI run for checking avx512, and oneDAL performance benchmarks of changes to function dispatching.

**Comment by icfaust** [comment]
Run with an avx512 build: http://intel-ci.intel.com/ef007c41-cb1f-f115-9514-a4bf010d0e2e failures due to un-related GPU issues.

**Comment by icfaust** [comment]
private CI failures due to unrelated GPU/dpc timeouts

**Comment by icfaust** [comment]
private CI run with last sklearnex master (includes _assert_all_finite tests coming from intel/scikit-learn-intelex#1759) http://intel-ci.intel.com/ef012d7e-a408-f166-adc8-a4bf010d0e2e

File: cpp/daal/src/data_management/finiteness_checker_avx2_impl.i [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __FINITENESS_CHECKER_AVX2_IMPL_I__
+#define __FINITENESS_CHECKER_AVX2_IMPL_I__
+
+template <>
+float sumWithAVX<float, avx2>(size_t n, const float * dataPtr)
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_avx2_impl.i:22]: I think further there might be non-AVX implementation of those functions. We already have 3 SIMD variants supported: SSE, AVX and Arm's SVE. There might be more in the future.

```suggestion
float sumWithSIMD<float, avx2>(size_t n, const float * dataPtr)
```

Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_avx2_impl.i:22]: Please add the description. Something like this.
```suggestion
/*
// Computes sum of the elements of input array of type `float` with AVX2 instructions.
*/
float sumWithAVX<float, avx2>(size_t n, const float * dataPtr)
```

File: cpp/daal/src/data_management/finiteness_checker_avx2_impl.i [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+template <>
+services::Status checkFinitenessInBlocks<avx2>(const float ** dataPtrs, bool inParallel, size_t nTotalBlocks, size_t nBlocksPerPtr, size_t nPerBlock,
+                                               size_t nSurplus, bool allowNaN, bool & finiteness)
+{
+    services::Status s;
+    constexpr size_t avx2RegisterLength = 256;
+    constexpr size_t numberOfBitsInByte = 8;
+    constexpr size_t nPerInstr          = avx2RegisterLength / (numberOfBitsInByte * sizeof(float));
+    services::internal::TArray<bool, avx2> notFiniteArr(nTotalBlocks);
+    bool * notFinitePtr = notFiniteArr.get();
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_avx2_impl.i:103]: This is just a FYI. No action is needed.

It might be better (in terms of performance) to have those kind of arrays as a 'thread local storage' (see daal::tls).

Having them as a regular arrays might lead to cache coherency issues when i-th thread updates i-th element of notFinitePtr array, and (i+1)-th thread updates (i+1)-th element of the same array. In this case both threads might need to synchronize the information in their caches as both i-th and (i+1)-th elements of the array might reside in both caches of i-th and (i+1)-th threads and can become unsynchronized after the update.

File: cpp/daal/src/data_management/finiteness_checker_cpu.cpp [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+    const size_t nCols                          = table.getNumberOfColumns();
+    NumericTableDictionaryPtr tableFeaturesDict = table.getDictionarySharedPtr();
+
+    daal::TlsMem<bool, cpu, services::internal::ScalableCalloc<bool, cpu> > tlsNotFinite(1);
+
+    daal::threader_for_break(nCols, nCols, [&](size_t i, bool & needBreak) {
+        bool * localNotFinite = tlsNotFinite.local();
+        DAAL_CHECK_MALLOC_THR(localNotFinite);
+
+        switch ((*tableFeaturesDict)[i].getIndexType())
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_cpu.cpp:199]: Don't we need something that checks for integer overflow for integer features here? How is it done in array API?

File: cpp/daal/src/data_management/finiteness_checker_cpu.cpp [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+    {
+        sum = computeSumSOA<cpu>(table, sumIsFinite, s);
+    }
+    else
+    {
+        ReadRows<DataType, cpu> dataBlock(table, 0, nRows);
+        DAAL_CHECK_BLOCK_STATUS(dataBlock);
+        const DataType * dataPtr = dataBlock.get();
+
+        sum = computeSum<DataType, cpu>(1, nElements, &dataPtr);
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_cpu.cpp:283]: Another FYI. No action is required.

It is better in terms of both performance and memory consumption to move ReadRows inside the parallel for that happens in computeSum.
Because ReadRows in some cases might copy the whole table and convert it -> +1 pass through the memory.
When ReadRows is made by blocks and is combined with the computational part it is usually more efficient.

File: cpp/daal/src/data_management/finiteness_checker_cpu.cpp [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+    #if (__CPUID__(DAAL_CPU) == __avx2__)
+
+        #include "finiteness_checker_avx2_impl.i"
+
+    #endif // __CPUID__(DAAL_CPU) == __avx2__
+
+#endif
+
+template <typename DataType, daal::CpuType cpu>
+services::Status allValuesAreFiniteImpl(NumericTable & table, bool allowNaN, bool * finiteness)
```
Comment (Vika-F) [inline cpp/daal/src/data_management/finiteness_checker_cpu.cpp:260]: Please add the description.

**Review by Vika-F — COMMENTED** [review]
Thank you for the restructuring effort.
The code became much more understandable.

My final comment is that 'AVX' needs to be replaced with something else like 'SIMD'. Otherwise the code is good to go.

**Comment by icfaust** [comment]
Rerun due to CI timeouts: http://intel-ci.intel.com/ef01f546-5586-f1d1-863c-a4bf010d0e2e

**Review by Vika-F — APPROVED** [review]
LGTM
